File: sdk/python/examples/09d_human_tool.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""Human Tool — LLM-initiated human interaction.

Unlike approval_required tools (09_human_in_the_loop.py) where humans gate
tool execution, ``human_tool`` lets the LLM **ask the human questions** at
any point.  The LLM decides when to call the tool, and the human's response
is returned as the tool output.

The tool is entirely server-side (Conductor HUMAN task) — no worker process
needed.  The server generates the response form and validation pipeline
automatically, so this works with any SDK language.

Demonstrates:
    - ``human_tool()`` for LLM-initiated human interaction
    - Mixing human tools with regular tools
    - The LLM using human input to make decisions

Requirements:
    - AGENTSPAN_SERVER_URL=http://localhost:8080/api
    - AGENTSPAN_LLM_MODEL (default: openai/gpt-4o-mini)
"""

from settings import settings

from agentspan.agents import Agent, AgentRuntime, EventType, human_tool, tool


@tool
def lookup_employee(name: str) -> dict:
    """Look up an employee by name and return their info."""
    employees = {
        "alice": {"name": "Alice Chen", "department": "Engineering", "level": "Senior"},
        "bob": {"name": "Bob Martinez", "department": "Sales", "level": "Manager"},
        "carol": {"name": "Carol Wu", "department": "Engineering", "level": "Staff"},
    }
    key = name.lower().split()[0]
    return employees.get(key, {"error": f"Employee '{name}' not found"})


@tool
def submit_ticket(title: str, priority: str, assignee: str) -> dict:
    """Submit an IT support ticket."""
    return {"ticket_id": "TKT-4821", "title": title, "priority": priority, "assignee": assignee}


ask_user = human_tool(
    name="ask_user",
    description="Ask the user a question when you need clarification or additional information.",
)

agent = Agent(
    name="it_support",
    model=settings.llm_model,
    tools=[lookup_employee, submit_ticket, ask_user],
    instructions=(
        "You are an IT support assistant. Help users create support tickets. "
        "Use lookup_employee to find employee info. "
        "If you need clarification about the issue or any details, use ask_user "
        "to ask the user directly. Always confirm the ticket details with the user "
        "before submitting."
    ),
)


if __name__ == "__main__":
    with AgentRuntime() as runtime:
        # Deploy to server. CLI alternative (recommended for CI/CD):
        #   agentspan deploy examples.09d_human_tool
        runtime.deploy(agent)
        runtime.serve(agent)

        # Quick test: uncomment below (and comment out serve) to run directly.
        # handle = runtime.start(agent, "I need to file a ticket for Alice about a laptop issue")
        # print(f"Workflow started: {handle.execution_id}\n")

        # for event in handle.stream():
        #     if event.type == EventType.THINKING:
        #         print(f"  [thinking] {event.content}")

        #     elif event.type == EventType.TOOL_CALL:
        #         print(f"  [tool_call] {event.tool_name}({event.args})")

        #     elif event.type == EventType.TOOL_RESULT:
        #         print(f"  [tool_result] {event.tool_name} -> {event.result}")

        #     elif event.type == EventType.WAITING:
        #         print("\n--- Human input required ---")
        #         response = input("  Your response: ").strip()
        #         handle.respond({"response": response})
        #         print()

        #     elif event.type == EventType.ERROR:
        #         print(f"  [error] {event.content}")

        #     elif event.type == EventType.DONE:
        #         print(f"\nResult: {event.output}")

File: sdk/typescript/examples/09d-human-tool.ts
/**
 * Human Tool — LLM-initiated human interaction.
 *
 * Unlike approvalRequired tools (09-human-in-the-loop.ts) where humans gate
 * tool execution, humanTool lets the LLM **ask the human questions** at
 * any point. The LLM decides when to call the tool, and the human's response
 * is returned as the tool output.
 *
 * The tool is entirely server-side (Conductor HUMAN task) — no worker process
 * needed. The server generates the response form and validation pipeline
 * automatically, so this works with any SDK language.
 *
 * Demonstrates:
 *   - humanTool() for LLM-initiated human interaction
 *   - Mixing human tools with regular tools
 *   - The LLM using human input to make decisions
 *
 * Requirements:
 *   - AGENTSPAN_SERVER_URL=http://localhost:6767/api
 *   - AGENTSPAN_LLM_MODEL (default: openai/gpt-4o-mini)
 */

import { z } from 'zod';
import { Agent, AgentRuntime, humanTool, tool } from '../src/index.js';
import type { AgentHandle } from '../src/index.js';
import { llmModel } from './settings.js';

const lookupEmployee = tool(
  async (args: { name: string }) => {
    const employees: Record<
      string,
      { name: string; department: string; level: string }
    > = {
      alice: {
        name: 'Alice Chen',
        department: 'Engineering',
        level: 'Senior',
      },
      bob: {
        name: 'Bob Martinez',
        department: 'Sales',
        level: 'Manager',
      },
      carol: {
        name: 'Carol Wu',
        department: 'Engineering',
        level: 'Staff',
      },
    };
    const key = args.name.toLowerCase().split(' ')[0];
    return employees[key] ?? { error: `Employee '${args.name}' not found` };
  },
  {
    name: 'lookup_employee',
    description: 'Look up an employee by name and return their info.',
    inputSchema: z.object({
      name: z.string().describe('Employee name to look up'),
    }),
  },
);

const submitTicket = tool(
  async (args: { title: string; priority: string; assignee: string }) => {
    return {
      ticket_id: 'TKT-4821',
      title: args.title,
      priority: args.priority,
      assignee: args.assignee,
    };
  },
  {
    name: 'submit_ticket',
    description: 'Submit an IT support ticket.',
    inputSchema: z.object({
      title: z.string().describe('Ticket title'),
      priority: z.string().describe('Priority level'),
      assignee: z.string().describe('Assignee name'),
    }),
  },
);

const askUser = humanTool({
  name: 'ask_user',
  description:
    'Ask the user a question when you need clarification or additional information.',
});

export const agent = new Agent({
  name: 'it_support',
  model: llmModel,
  tools: [lookupEmployee, submitTicket, askUser],
  instructions:
    'You are an IT support assistant. Help users create support tickets. ' +
    'Use lookup_employee to find employee info. ' +
    'If you need clarification about the issue or any details, use ask_user ' +
    'to ask the user directly. Always confirm the ticket details with the user ' +
    'before submitting.',
});

// Only run when executed directly (not when imported for discovery)
if (process.argv[1]?.endsWith('09d-human-tool.ts') || process.argv[1]?.endsWith('09d-human-tool.js')) {
  const runtime = new AgentRuntime();
  try {
    const result = await runtime.run(
      agent,
      'Look up Alice Chen and summarize her department and level.',
    );
    result.printResult();

    // Production pattern:
    // 1. Deploy once during CI/CD:
    // await runtime.deploy(agent);
    // CLI alternative:
    // agentspan deploy --package sdk/typescript/examples --agents it_support
    //
    // 2. In a separate long-lived worker process:
    // await runtime.serve(agent);

    // Interactive human-tool alternative:
    // const handle: AgentHandle = await runtime.start(
    //   agent,
    //   'I need to file a ticket for Alice about a laptop issue',
    // );
    // console.log(`Execution started: ${handle.executionId}\n`);
    // for await (const event of handle.stream()) {
    //   if (event.type === 'waiting') {
    //     await handle.respond({
    //       response:
    //         'The laptop screen is flickering, high priority please',
    //     });
    //   }
    // }
  } finally {
    await runtime.shutdown();
  }
}
